firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Power/Simulator.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Simulator.cpp
+ *
+ *  Created on: May 30, 2021
+ *      Author: Rohan
+ *
+ *  Simulator for power management
+ */
+
+#include "Pufferfish/Driver/Power/Simulator.h"
+
+#include <cmath>
+#include <random>
+
+#include "Pufferfish/Util/Timeouts.h"
+
+namespace Pufferfish::Driver::Power {
+
+// NOLINTNEXTLINE(cert-msc51-cpp)
+std::ranlux24_base prng;
+// NOLINTNEXTLINE(readability-magic-numbers)
+std::uniform_int_distribution<int> uniform_int(1, 100);
+static constexpr float uniform_width = 1;
+std::uniform_real_distribution<float> uniform_centered(-uniform_width / 2, uniform_width / 2);
+
+void Simulator::input_clock(uint32_t current_time) {
+  if (initial_time_ == 0) {
+    initial_time_ = current_time;
+  }
+  if (update_needed()) {
+    previous_time_ = current_time_;
+  }
+  current_time_ = current_time - initial_time_;
+}
+
+void Simulator::transform(uint32_t current_time, PowerManagement &power_management) {
+  input_clock(current_time);
+  if (charging_) {
+    transform_charge(power_management);
+  } else {
+    initial_time_ = current_time;
+    transform_discharge(power_management);
+  }
+}
+
+void Simulator::transform_charge(PowerManagement &power_management) {
+  if (!update_needed()) {
+    return;
+  }
+  power_management.charging = true;
+  power_management.power_left +=
+      (1 + (power_responsiveness * uniform_centered(prng))) / uniform_int(prng);
```

Comment:
This formula is seems more convoluted than necessary. If you're simply trying to add a random percentage to `power_left`, it would be clearer to just define a uniform random distribution ` std::uniform_real_distribution<float> uniform_percentage(min_increase_per_step, max_increase_per_step)` (and you can just delete `uniform_int` and `uniform_centered`) and add `uniform_percentage(prng)` to `power_left`.

Reply:
I'm now simply incrementing from 0 to 100 every second and decrementing on discharge, works fine and we can get rid of random distributions and randomness in results too, I'm resolving this!